HPCC-9259 Add JSON support to ESP

mayx opened the pull request:
- Create a pull parser interface that will be shared by new parser class
  and original XmlPullParser class
- Implement the pull parser class using jlib JSON and XML pull parsers
- Serialize directly from rpc response to JSON text
- Add JSON test to ESP service form page

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

mayx commented:
@afishbeck please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 18d53536fd62976f50e4c458d6508ddef0144cfd
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |787 |787 |0 |
|test (thor) |718 |718 |0 |
|test (roxie) |831 |831 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  177 sec (00:02:57) |   61 sec (00:01:01) |    9 sec (00:00:09) |   29 sec (00:00:29) | 1266 sec (00:21:06) |   19 sec (00:00:19) | 1578 sec (00:26:18) |

afishbeck reviewed (approved):
@mayx a few trivial comments, but otherwise looks good.

mayx committed:
HPCC-9259 Add JSON support to ESP

- Create a pull parser interface that will be shared by new parser class
  and original XmlPullParser class
- Implement the pull parser class using jlib JSON and XML pull parsers
- Serialize directly from rpc response to JSON text
- Add JSON test to ESP service form page

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

mayx commented on esp/bindings/SOAP/Platform/soapbind.cpp:
```diff
@@ -108,35 +110,66 @@ static CSoapFault* makeSoapFault(CHttpRequest* request, IMultiException* me, con
+        {
+            StringBuffer ns;
+            soapFault.setown(makeSoapFault(request,mex, generateNamespace(*request->queryContext(), request, request->queryServiceName(), request->queryServiceMethod(), ns).str()));
+            //SetHTTPErrorStatus(mex->errorCode(),response);
```
It was set to be always 500 for SOAP, don't know exactly the reason but I'll leave it that way. For JSON I've now changed it to use the SetHTTPErrorStatus function to reflect the error status more accurately.

mayx replied:
(The commented code isn't new, it's labeled as new because of other related and indentation changes)

afishbeck replied:
I think SOAPFaults may be specified to always be 500.  REST should reflect appropriate HTTP status codes.

mayx commented:
@afishbeck changes made per suggestions, please review again...

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: fb81cb5923d5f85a2b2a62944f4265d7ecd41b6c
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |790 |790 |0 |
|test (thor) |721 |721 |0 |
|test (roxie) |861 |861 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  173 sec (00:02:53) |   62 sec (00:01:02) |    7 sec (00:00:07) |   27 sec (00:00:27) | 1459 sec (00:24:19) |   18 sec (00:00:18) | 1755 sec (00:29:15) |

afishbeck reviewed (approved):
@mayx looks good.
